### PR TITLE
Image Classification Python Task Implementation

### PR DESCRIPTION
**Pull request by kinaryml**
- Added the implementation for the image classifier and associated tests.

### CODE REVIEW

File: mediapipe/tasks/python/vision/image_classifier.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            ':'.join([_IMAGE_TAG, _IMAGE_OUT_STREAM_NAME])
+        ],
+        task_options=options)
+    return cls(
+        task_info.generate_graph_config(
+            enable_flow_limiting=options.running_mode ==
+            _RunningMode.LIVE_STREAM), options.running_mode,
+        packets_callback if options.result_callback else None)
+
+  # TODO: Create an Image class for MediaPipe Tasks.
```
Comment (jiuqiant) [inline mediapipe/tasks/python/vision/image_classifier.py:159]: Remove this TODO.

File: mediapipe/tasks/python/vision/image_classifier.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    Returns:
+      A classification result object that contains a list of classifications.
+
+    Raises:
+      ValueError: If any of the input arguments is invalid.
+      RuntimeError: If image classification failed to run.
+    """
+    output_packets = self._process_video_data({
+      _IMAGE_IN_STREAM_NAME:
+        packet_creator.create_image(image).at(timestamp_ms)
```
Comment (jiuqiant) [inline mediapipe/tasks/python/vision/image_classifier.py:210]: I believe this needs to be multiplied by 1000 since the timestamps of mediapipe packets are in microseconds.

File: mediapipe/tasks/python/vision/image_classifier.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    })
+    classification_result_proto = packet_getter.get_proto(
+      output_packets[_CLASSIFICATION_RESULT_OUT_STREAM_NAME])
+
+    return classifications_module.ClassificationResult([
+        classifications_module.Classifications.create_from_pb2(classification)
+        for classification in classification_result_proto.classifications
+    ])
```
Comment (jiuqiant) [inline mediapipe/tasks/python/vision/image_classifier.py:218]: If possible, please add the classify_async api into this PR and update the unit tests.

File: mediapipe/tasks/python/test/vision/image_classifier_test.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+_SCORE_THRESHOLD = 0.5
+_MAX_RESULTS = 3
+
+
+class ModelFileType(enum.Enum):
+  FILE_CONTENT = 1
+  FILE_NAME = 2
+
+
+class ImageClassifierTest(parameterized.TestCase):
```
Comment (jiuqiant) [inline mediapipe/tasks/python/test/vision/image_classifier_test.py:88]: Better to test more corner cases such as the invalid configurations. Please take a look at the cc test and the java test for more information.

**Review by jiuqiant — COMMENTED** [review]
Thanks for making the image classifier api. I left a few comments. Please address them if possible. Thanks!

File: mediapipe/tasks/python/vision/image_classifier.py (hunk 3 above)
Comment (kinaryml) [reply mediapipe/tasks/python/vision/image_classifier.py:218]: @jiuqiant I encountered the following while trying to run tests for the `classify_async` API.

```
F20221008 05:19:47.966590    64 packet.cc:138] Check failed: holder_ != nullptr The packet is empty.
```

I'll include it in this PR with code the code commented out so you try reproducing the issue. Everything else seems to run fine.

File: mediapipe/tasks/python/vision/image_classifier.py (hunk 1 above)
Comment (kinaryml) [reply mediapipe/tasks/python/vision/image_classifier.py]: Removed it

File: mediapipe/tasks/python/vision/image_classifier.py (hunk 2 above)
Comment (kinaryml) [reply mediapipe/tasks/python/vision/image_classifier.py]: Added a new `_MICRO_SECONDS_PER_MILLISECOND` constant

File: mediapipe/tasks/python/test/vision/image_classifier_test.py (hunk 4 above)
Comment (kinaryml) [reply mediapipe/tasks/python/test/vision/image_classifier_test.py:102]: @jiuqiant I've added all the remaining tests now. Though I happened to run into an issue with the final test for the `classify_async` API. PTAL after uncommenting the code and running the tests again.

**Comment by kinaryml** [comment]
I've added support to accept an `roi` (region of interest) parameter to the `classify` and `classify_for_video` APIs now.

File: mediapipe/tasks/python/test/vision/image_classifier_test.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+"""Tests for image classifier."""
+
+import enum
+from unittest import mock
+
+import numpy as np
+from absl.testing import absltest
+from absl.testing import parameterized
+
+from mediapipe.python._framework_bindings import image as image_module
```
Comment (lu-wang-g) [inline mediapipe/tasks/python/test/vision/image_classifier_test.py:23]: @jiuqiant Should we move image to another namespace, i.e. mediapipe.framework.image? It's probably easier to make it an alias of mediapipe.python._framework_bindings.image.

**Comment by lu-wang-g** [comment]
Looped in Maxime to confirm the namespace, options, I/O containers, etc.

File: mediapipe/tasks/python/vision/image_classifier.py (hunk 3 above)
Comment (jiuqiant) [reply mediapipe/tasks/python/vision/image_classifier.py:238]: MediaPipe returns empty packets if there is nothing detected. Probably we need to check if the packet is empty, do something like:

```
if output_packets[_CLASSIFICATION_RESULT_OUT_STREAM_NAME].is_empty():
  # return empty classification result
else:
  # make a classification result from packet.
```

File: mediapipe/tasks/python/vision/image_classifier.py [hunk 6]
```diff
@@ -137,12 +147,14 @@ def packets_callback(output_packets: Mapping[str, packet_module.Packet]):
       ])
       image = packet_getter.get_image(output_packets[_IMAGE_OUT_STREAM_NAME])
```
Comment (jiuqiant) [inline mediapipe/tasks/python/vision/image_classifier.py:148]: Do a similar check first: https://github.com/google/mediapipe/blob/master/mediapipe/tasks/python/vision/object_detector.py#L147

File: mediapipe/tasks/python/test/vision/image_classifier_test.py (hunk 5 above)
Comment (jiuqiant) [reply mediapipe/tasks/python/test/vision/image_classifier_test.py:23]: You mean reexporting mediapipe.python._framework_bindings.image as mediapipe.framework.image?

File: mediapipe/tasks/python/test/vision/image_classifier_test.py (hunk 4 above)
Comment (jiuqiant) [reply mediapipe/tasks/python/test/vision/image_classifier_test.py:124]: Can you copy and paste the error log? I can take a quick look.

File: mediapipe/tasks/python/test/vision/image_classifier_test.py (hunk 5 above)
Comment (lu-wang-g) [reply mediapipe/tasks/python/test/vision/image_classifier_test.py:23]: Correct. Having `_framework_bindings` in the import path is wired ...

This can be done by simply adding the file, mediapipe/python/framework/__init__.py, and put the following code inside:
```
from mediapipe.python._framework_bindings import image as image_module

image = image_module
```

File: mediapipe/tasks/python/test/vision/image_classifier_test.py (hunk 4 above)
Comment (kinaryml) [reply mediapipe/tasks/python/test/vision/image_classifier_test.py:127]: @jiuqiant I've fixed it now.

File: mediapipe/tasks/python/vision/image_classifier.py (hunk 3 above)
Comment (kinaryml) [reply mediapipe/tasks/python/vision/image_classifier.py:240]: @jiuqiant Thanks. I was missing this check and now the API runs fine after adding it.

File: mediapipe/tasks/python/test/vision/image_classifier_test.py [hunk 7]
```diff
@@ -492,12 +500,6 @@ def check_result(result: _ClassificationResult, unused_output_image: _Image,
       classifier_options=classifier_options,
       result_callback=check_result)
     classifier = _ImageClassifier.create_from_options(options)
```
[the remark this answers is not in the document]
Comment (kinaryml) [reply mediapipe/tasks/python/test/vision/image_classifier_test.py]: @jiuqiant I've changed it now.

**Review by schmidt-sebastian — APPROVED** [review]
